src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs

```diff
@@ -0,0 +1,354 @@
+            {
+                _logger.LogError("Error decoding compound code");
+                throw;
+            }
+
+            Uri callbackUrl = new Uri(
+                Request.Scheme + "://" + Request.Host + "/AadProxy/callback/" +
+                Base64UrlEncoder.Encode(redirectUri.ToString()));
+
+            // TODO: Deal with client secret in basic auth header
+            var content = new FormUrlEncodedContent(
+                new[]
+                {
+                    new KeyValuePair<string, string>("grant_type", grantType),
+                    new KeyValuePair<string, string>("code", code),
+                    new KeyValuePair<string, string>("redirect_uri", callbackUrl.ToString()),
+                    new KeyValuePair<string, string>("client_id", clientId),
+                    new KeyValuePair<string, string>("client_secret", clientSecret),
+                });
+
+            var response = await client.PostAsync(new Uri(_aadTokenEndpoint), content);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                return new ContentResult()
+                {
+                    Content = await response.Content.ReadAsStringAsync(),
+                    StatusCode = (int)response.StatusCode,
+                    ContentType = "application/json",
+                };
+            }
+
+            var tokenResponse = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            // Handle fields passed through launch context
+            foreach (var launchField in _launchContextFields)
+            {
+                if (decodedCompoundCode[launchField] != null)
+                {
+                    tokenResponse[launchField] = decodedCompoundCode[launchField];
+                }
+            }
+
+            tokenResponse["client_id"] = clientId;
+
+            // Replace fully qualifies scopes with short scopes and replace $
+            string[] scopes = tokenResponse["scope"].ToString().Split(' ');
+            StringBuilder scopesBuilder = new StringBuilder();
+
+            foreach (var s in scopes)
+            {
+                if (IsAbsoluteUrl(s))
+                {
+                    Uri scopeUri = new Uri(s);
+                    scopesBuilder.Append($"{scopeUri.Segments.Last().Replace('$', '/')} ");
+                }
+                else
+                {
+                    scopesBuilder.Append($"{s.Replace('$', '/')} ");
+                }
+            }
+
+            tokenResponse["scope"] = scopesBuilder.ToString().TrimEnd(' ');
+
+            return new ContentResult()
+            {
+                Content = tokenResponse.ToString(Newtonsoft.Json.Formatting.None),
+                StatusCode = (int)response.StatusCode,
+                ContentType = "application/json",
+            };
+        }
+
+        private static bool IsAbsoluteUrl(string url)
+        {
+            Uri result;
+            return Uri.TryCreate(url, UriKind.Absolute, out result);
```

Comment:
Syntactic sugar: you could write `return Uri.TryCreate(url, UriKind.Absolute, out _);`